payments-core/src/main/java/com/stripe/android/payments/core/authentication/DefaultIntentAuthenticatorRegistry.kt

```diff
@@ -0,0 +1,147 @@
+import com.stripe.android.StripePaymentController
+import com.stripe.android.model.StripeIntent
+import com.stripe.android.networking.AnalyticsRequestExecutor
+import com.stripe.android.networking.AnalyticsRequestFactory
+import com.stripe.android.networking.StripeRepository
+import com.stripe.android.payments.PaymentFlowResult
+import com.stripe.android.payments.core.injection.AuthenticationModule
+import com.stripe.android.payments.core.injection.DaggerAuthenticationComponent
+import com.stripe.android.stripe3ds2.service.StripeThreeDs2Service
+import com.stripe.android.stripe3ds2.transaction.MessageVersionRegistry
+import com.stripe.android.view.AuthActivityStarter
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Default registry to provide look ups for [IntentAuthenticator].
+ * Should be only accessed through [DefaultIntentAuthenticatorRegistry.getInstance].
+ */
+internal class DefaultIntentAuthenticatorRegistry @Inject internal constructor() :
+    IntentAuthenticatorRegistry {
+
+    // default authenticators always packaged with the SDK
+    // TODO(ccen) register a singleton instance of each authenticator with registration API
+    @Inject
+    lateinit var webIntentAuthenticator: WebIntentAuthenticator
+
+    @Inject
+    lateinit var noOpIntentAuthenticator: NoOpIntentAuthenticator
+
+    // authenticators requiring a 3p SDK
+    // TODO(ccen) move them to a dedicated module
+    @Inject
+    lateinit var stripe3DS2Authenticator: Stripe3DS2Authenticator
+
+    override fun lookUp(stripeIntent: StripeIntent): IntentAuthenticator {
+        if (!stripeIntent.requiresAction()) {
+            return noOpIntentAuthenticator
+        }
+
+        when (val nextActionData = stripeIntent.nextActionData) {
+            is StripeIntent.NextActionData.SdkData.Use3DS2 -> {
+                return stripe3DS2Authenticator
+            }
+            is StripeIntent.NextActionData.SdkData.Use3DS1 -> {
+                // can only triggered when `use_stripe_sdk=true`
+                return webIntentAuthenticator
+            }
+            is StripeIntent.NextActionData.RedirectToUrl -> {
+                // can only triggered when `use_stripe_sdk=false`
+                return webIntentAuthenticator
+            }
+            is StripeIntent.NextActionData.AlipayRedirect -> {
+                return webIntentAuthenticator
+            }
+            is StripeIntent.NextActionData.DisplayOxxoDetails -> {
+                return webIntentAuthenticator.takeIf { nextActionData.hostedVoucherUrl != null }
+                    ?: noOpIntentAuthenticator
+            }
+            else -> return noOpIntentAuthenticator
+        }
+    }
+
+    companion object {
+        // Holding this single instance would in turn holds DaggerAuthenticationComponent instance,
+        // which keeps dagger injection graph live.
+        @Volatile
+        private var INSTANCE: IntentAuthenticatorRegistry? = null
```

Comment:
Why is `@Volatile` needed?

Reply:
the [double checked locking](https://en.wikipedia.org/wiki/Double-checked_locking) is used to guard race conditions that might access the method in multi threads, I think `@Volatile` makes sure there is only one copy of instances across multi threads. this snippet is taken from https://stackoverflow.com/a/45943282/802372.

On a second thought, I think we shouldn't use a singleton instance here, just reverted this to a new instance creation each time, the reason is as follows-

* **before** the change, a`Stripe` instance holds a `StripePaymentController` instance, which holds logics of the registry we extracted here

* **after** the change, a`Stripe` instance holds a `StripePaymentController` instance, which holds a `IntentAuthenticatorRegistry` instance 
  * If we make `IntentAuthenticatorRegistry` a singleton, it will be shared if user creates multiple `Stripe` instances, though this doesn't seem to break anything as the registry is stateless and can be shared, I feel like we should keep the logic in sync before the change